sound/soc/sof/control.c

```diff
@@ -82,8 +82,10 @@ int snd_sof_volume_put(struct snd_kcontrol *kcontrol,
 		(struct soc_mixer_control *)kcontrol->private_value;
```

Comment:
generic comment: I think we are confusing two layers.
a) the IPC failed due to scheduling/timeout issues
b) the data provided to the DSP is invalid.

This PR does not explain how these two different cases are handled. 
In case a) it's to be hoped that the IPC timeout will not happen again on a resume, and this was a transient error.
In case b) the setting will never work unless the application changes the values. It's not a matter of IPC proper but of bad data being rejected.

And I am personally not aware of any specific error-handling policy for controls, not sure either how the two cases should be handled.

Reply:
Currently there's no separated handling of -EBUSY (The request is valid but DSP cannot do it now), -ENOMEM (DSP cannot allocate now the RAM, but in theory could have success another time), and -EINVAL (illegal request, will be always rejected). All of these are treated as error. I left with this PR the behavior as is. It would be quite risk to place the request into scontrols cache for state restore. It's quite possible that they would fail in next attempt too by PM restore. That would be a way to brick a feature in a running SOF system. So without more detailed information it's better to keep previous accepted request in cache. This PR protects illegal request to overwrite a previous good request in PM cache.

The DSP responds in addition to 0/-EINVAL/... etc. with random data words. I think it should instead respond with an empty message if there is no additional data to respond with. But since kernel needs to be compatible with previous ABI versions the kernel needs to be made immune against corrupt data from DSP. This patch implements that by no letting DSP response to be written over to scontrols cache.

So, I'll rebase today and improve commit text.

Reply:
@plbossart @ranj063 The previous PR version implemented also in lower level IPC code separation of request and response buffers. This again doesn't use it. Both protecting the scontrols cache and separating the flows could be done if you prefer. Though with current IPC needs it does not seem to be necessary (dual kzallocs and data copies done). Please let me know your preference for that as well to get this PR technically into best possible shape.